### PR TITLE
kudu: use in-list predicate if ranges are all single values

### PR DESCRIPTION
**Pull request by lzeiming**
optimize for query like:
select * from kudu_table where str_column in ( 'hello', 'world');

### CODE REVIEW

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java [hunk 1]
```diff
@@ -500,20 +500,31 @@ else if (domain.isSingleValue()) {
                     }
```
Comment (kokosing) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: I would rewrite cmt msg to:
```
Use in-list predicate if ranges are all single values in Kudu

That optimizes queries like:
SELECT * FROM kudu_table WHERE str_column IN ('hello', 'world');
```

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java [hunk 2]
```diff
@@ -500,20 +500,31 @@ else if (domain.isSingleValue()) {
...
                     else if (valueSet instanceof SortedRangeSet) {
```
Comment (kokosing) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: Do you have an example of speedup you got with this change?

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java [hunk 3]
```diff
@@ -500,20 +500,31 @@ else if (domain.isSingleValue()) {
...
                         Ranges ranges = ((SortedRangeSet) valueSet).getRanges();
-                        Range span = ranges.getSpan();
-                        Marker low = span.getLow();
-                        if (!low.isLowerUnbounded()) {
-                            KuduPredicate.ComparisonOp op = (low.getBound() == Marker.Bound.ABOVE)
-                                    ? KuduPredicate.ComparisonOp.GREATER : KuduPredicate.ComparisonOp.GREATER_EQUAL;
-                            KuduPredicate predicate = createComparisonPredicate(columnSchema, op, low.getValue());
+                        List<Range> rangeList = ranges.getOrderedRanges();
+                        if (rangeList.stream().allMatch(Range::isSingleValue)) {
+                            io.prestosql.spi.type.Type type = TypeHelper.fromKuduColumn(columnSchema);
```
Comment (kokosing) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: can you use import?

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java [hunk 4]
```diff
@@ -500,20 +500,31 @@ else if (domain.isSingleValue()) {
                     }
                     else if (valueSet instanceof SortedRangeSet) {
                         Ranges ranges = ((SortedRangeSet) valueSet).getRanges();
```
Comment (kokosing) [inline presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: Do you have unit test coverage for this queries? Alternatively can you run `io.prestosql.tests.AbstractTestQueries` for Kudu connectors? I mean can you create `TestKuduQueries` that would extend from `AbstractTestQueries`? I wonder how you change would improve `io.prestosql.tests.AbstractTestQueries#testLargeIn` test.

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java (hunk 3 above)
Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: org.apache.kudu.Type is imported, So qualified name should be used here.

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java (hunk 2 above)
Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: Itest with a table with 400 million rows, queries like SELECT * FROM kudu_table WHERE str_column IN (...); could be reduced from 47 seconds to 3 secends.  Rows presto scaned is reduced from hundred millions to  the size of in-list

File: presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java (hunk 4 above)
Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: I tried testLargeIn for Kudu and found no improvement for this test case.

original:
2019-12-12T23:51:45.345-0600 INFO Test io.prestosql.plugin.kudu.TestKuduQueries::init took 1.20m
2019-12-12T23:52:03.267-0600 INFO FINISHED in presto: 17.79s, h2: 83.62ms, total: 17.88s
2019-12-12T23:52:03.302-0600 INFO TIMELINE: Query 20191213_055145_00008_4rf83 :: Transaction:[9f1de477-fdb0-4552-af21-c507e6724131] :: elapsed 17622ms :: planning 15850ms :: waiting 52ms :: scheduling 1516ms :: running 154ms :: finishing 102ms :: begin 2019-12-12T23:51:45.491-06:00 :: end 2019-12-12T23:52:03.113-06:00
2019-12-12T23:52:19.610-0600 INFO FINISHED in presto: 16.30s, h2: 37.11ms, total: 16.33s
2019-12-12T23:52:19.638-0600 INFO TIMELINE: Query 20191213_055203_00009_4rf83 :: Transaction:[6aeff91f-6821-4c21-bc8f-e79895fc69eb] :: elapsed 16232ms :: planning 14874ms :: waiting 25ms :: scheduling 890ms :: running 423ms :: finishing 45ms :: begin 2019-12-12T23:52:03.308-06:00 :: end 2019-12-12T23:52:19.540-06:00
2019-12-12T23:52:34.395-0600 INFO TIMELINE: Query 20191213_055219_00010_4rf83 :: Transaction:[a2d5927f-e976-4779-8107-02af9d4a537c] :: elapsed 14693ms :: planning 13280ms :: waiting 80ms :: scheduling 840ms :: running 560ms :: finishing 13ms :: begin 2019-12-12T23:52:19.680-06:00 :: end 2019-12-12T23:52:34.373-06:00
2019-12-12T23:52:36.668-0600 INFO FINISHED in presto: 14.73s, h2: 2.29s, total: 17.02s
2019-12-12T23:52:53.196-0600 INFO TIMELINE: Query 20191213_055236_00011_4rf83 :: Transaction:[5c622cf4-16c5-4206-8fc0-0ab3fcf61a97] :: elapsed 16427ms :: planning 15363ms :: waiting 26ms :: scheduling 802ms :: running 220ms :: finishing 42ms :: begin 2019-12-12T23:52:36.693-06:00 :: end 2019-12-12T23:52:53.120-06:00
2019-12-12T23:52:54.011-0600 INFO Major GC: application 76837ms, stopped 455ms: 667.37MB -> 635.21MB
2019-12-12T23:52:54.011-0600 INFO Major GC: application 76837ms, stopped 455ms: 667.37MB -> 635.21MB
2019-12-12T23:52:54.011-0600 INFO Major GC: application 76837ms, stopped 455ms: 667.37MB -> 635.21MB
2019-12-12T23:52:55.871-0600 INFO FINISHED in presto: 16.49s, h2: 2.71s, total: 19.20s
2019-12-12T23:52:56.963-0600 INFO FINISHED in presto: 1.04s, h2: 14.36ms, total: 1.06s
2019-12-12T23:52:56.964-0600 INFO TIMELINE: Query 20191213_055255_00012_4rf83 :: Transaction:[367d6e16-17b7-4210-91ef-ef3d4cc21793] :: elapsed 965ms :: planning 849ms :: waiting 39ms :: scheduling 80ms :: running 29ms :: finishing 7ms :: begin 2019-12-12T23:52:55.979-06:00 :: end 2019-12-12T23:52:56.944-06:00
2019-12-12T23:52:57.792-0600 INFO TIMELINE: Query 20191213_055256_00013_4rf83 :: Transaction:[49b51a39-8a46-4022-b9d6-a61df2bdff7b] :: elapsed 761ms :: planning 697ms :: waiting 30ms :: scheduling 45ms :: running 16ms :: finishing 3ms :: begin 2019-12-12T23:52:57.027-06:00 :: end 2019-12-12T23:52:57.788-06:00
2019-12-12T23:52:57.792-0600 INFO Test io.prestosql.plugin.kudu.TestKuduQueries::testLargeIn took 1.21m


patched:
2019-12-13T01:56:34.638-0600 INFO Test io.prestosql.plugin.kudu.TestKuduQueries::init took 50.27s
2019-12-13T01:56:52.268-0600 INFO TIMELINE: Query 20191213_075634_00008_qt2ba :: Transaction:[a4fbac16-ee45-4af8-a02a-b6aec56fefe5] :: elapsed 17361ms :: planning 15828ms :: waiting 43ms :: scheduling 1164ms :: running 324ms :: finishing 45ms :: begin 2019-12-13T01:56:34.762-06:00 :: end 2019-12-13T01:56:52.123-06:00
2019-12-13T01:56:52.279-0600 INFO FINISHED in presto: 17.48s, h2: 111.52ms, total: 17.60s
2019-12-13T01:57:07.451-0600 INFO FINISHED in presto: 15.13s, h2: 31.71ms, total: 15.16s
2019-12-13T01:57:07.476-0600 INFO TIMELINE: Query 20191213_075652_00009_qt2ba :: Transaction:[bc70b040-e4e0-4cb3-84b7-263c700f62ea] :: elapsed 15072ms :: planning 13961ms :: waiting 29ms :: scheduling 844ms :: running 230ms :: finishing 37ms :: begin 2019-12-13T01:56:52.313-06:00 :: end 2019-12-13T01:57:07.385-06:00
2019-12-13T01:57:14.013-0600 INFO Major GC: application 55419ms, stopped 471ms: 632.53MB -> 607.91MB
2019-12-13T01:57:14.013-0600 INFO Major GC: application 55419ms, stopped 471ms: 632.53MB -> 607.91MB
2019-12-13T01:57:14.013-0600 INFO Major GC: application 55419ms, stopped 471ms: 632.53MB -> 607.91MB
2019-12-13T01:57:22.947-0600 INFO TIMELINE: Query 20191213_075707_00010_qt2ba :: Transaction:[5ed1a36c-b5f3-46c7-b237-376fe783da1a] :: elapsed 15411ms :: planning 14242ms :: waiting 69ms :: scheduling 752ms :: running 407ms :: finishing 10ms :: begin 2019-12-13T01:57:07.521-06:00 :: end 2019-12-13T01:57:22.932-06:00
2019-12-13T01:57:25.209-0600 INFO FINISHED in presto: 15.45s, h2: 2.27s, total: 17.72s
2019-12-13T01:57:43.134-0600 INFO TIMELINE: Query 20191213_075725_00011_qt2ba :: Transaction:[d9da604b-b419-4e81-9273-73a8b25383a5] :: elapsed 17816ms :: planning 15352ms :: waiting 23ms :: scheduling 882ms :: running 1540ms :: finishing 42ms :: begin 2019-12-13T01:57:25.244-06:00 :: end 2019-12-13T01:57:43.060-06:00
2019-12-13T01:57:45.366-0600 INFO FINISHED in presto: 17.88s, h2: 2.27s, total: 20.15s
2019-12-13T01:57:46.532-0600 INFO TIMELINE: Query 20191213_075745_00012_qt2ba :: Transaction:[8f671759-b86e-42f1-88d3-90eb649ca419] :: elapsed 1022ms :: planning 914ms :: waiting 44ms :: scheduling 74ms :: running 28ms :: finishing 6ms :: begin 2019-12-13T01:57:45.500-06:00 :: end 2019-12-13T01:57:46.522-06:00
2019-12-13T01:57:46.539-0600 INFO FINISHED in presto: 1.13s, h2: 9.85ms, total: 1.14s
2019-12-13T01:57:47.398-0600 INFO TIMELINE: Query 20191213_075746_00013_qt2ba :: Transaction:[89dcaeee-1d5a-406c-9554-5de0f24aa69c] :: elapsed 750ms :: planning 703ms :: waiting 22ms :: scheduling 35ms :: running 10ms :: finishing 2ms :: begin 2019-12-13T01:57:46.643-06:00 :: end 2019-12-13T01:57:47.393-06:00
2019-12-13T01:57:47.399-0600 INFO Test io.prestosql.plugin.kudu.TestKuduQueries::testLargeIn took 1.21m

Comment (kokosing) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: Is LargeIn too large?

Comment (kokosing) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: Are you going to add `TestKuduQueries` to this pull request (as separate commit)?

Comment (kokosing) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: > Is LargeIn too large?

Please give it a shoot and try to understand why it is not working. Maybe there is a bug.

Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: may be the in-list is continuous, so the original push predicate for this case is optimal

Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: I test with this:
        assertQuery("SELECT orderkey FROM orders WHERE orderkey IN (1, 60000)");
improvement is not obvious.
 I think this patch only improve on large amount of rows and when the in-list is sparse and the gap between min and max is large.

orignal:
2019-12-13T06:04:16.529-0600 INFO FINISHED in presto: 6.13s, h2: 45.82ms, total: 6.18s
2019-12-13T06:04:16.618-0600 INFO TIMELINE: Query 20191213_120411_00000_m33cy :: Transaction:[7bd235d2-b0c4-425e-833a-012d653318a3] :: elapsed 5131ms :: planning 672ms :: waiting 2050ms :: scheduling 2889ms :: running 1500ms :: finishing 70ms :: begin 2019-12-13T06:04:11.336-06:00 :: end 2019-12-13T06:04:16.467-06:00

patched:
2019-12-13T06:05:38.488-0600 INFO FINISHED in presto: 5.99s, h2: 40.67ms, total: 6.03s
2019-12-13T06:05:38.646-0600 INFO TIMELINE: Query 20191213_120533_00000_uj7ww :: Transaction:[4067001d-9a36-4de6-a14a-c2337df00725] :: elapsed 4901ms :: planning 659ms :: waiting 1929ms :: scheduling 2784ms :: running 1372ms :: finishing 86ms :: begin 2019-12-13T06:05:33.530-06:00 :: end 2019-12-13T06:05:38.431-06:00

Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: There are hundreds of testcases in AbstractTestQueries, I am not sure all of them can be passed with kudu. So I do not include TestKuduQueries.

Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: There are hundreds of testcases in AbstractTestQueries, I am not sure all of them can be passed with kudu. So I do not include TestKuduQueries.

Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: There are hundreds of testcases in AbstractTestQueries, I am not sure all of them can be passed with kudu. So I do not include TestKuduQueries.

Comment (lzeiming) [reply presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduClientSession.java]: There are hundreds of testcases in AbstractTestQueries, I am not sure all of them can be passed with kudu. So I do not include TestKuduQueries.

**Comment by kokosing** [comment]
Merged, thanks!

**Pull request merged**